### DISCUSSION TITLE
Running ragna ui in daemon mode with configurable number of workers

### DISCUSSION THREAD

**Discussion by NetDevAutomate** [Quansight/ragna#208] [Q&A] [answered]
Is there any way to run ragna ui in daemon mode with a configurable number of workers?

I could run it using nohup once I know how to set the number of workers in the config file, but it would be nice to have a cross platform daemon mode.

**Comment by pmeier** [comment]
> Is there any way to run ragna ui in daemon mode

I don't understand the question. What do you want to achieve. What does "daemon" mean for you in this context?

> configurable number of workers?

Yes, of course. You just need to start the worker and API manually, i.e.

```bash
$ ragna worker --num-threads 16
$ ragna api --no-start-worker
$ ragna ui --no-start-api
```

By default `ragna ui` will start the API if it doesn't find it running and `ragna api` in turn will start a worker with a single thread.

**Reply by NetDevAutomate** [reply]
When I mean daemon I mean that ragna will run in the background, so if the the terminal session dies the process won't.

`ragna worker --num-threads 16

INFO:huey.consumer:Huey consumer started with 16 thread, PID 2963013 at 2023-11-19 10:26:33.106167
INFO:huey.consumer:Scheduler runs every 1 second(s).
INFO:huey.consumer:Periodic tasks are enabled.
INFO:huey.consumer:The following commands are available:
+ ragna.core._queue._Task`

As ragna runs as a forground process, I can't then run further cli commands in the same terminal.

Have I missed something obvious?

Apologies for all the questions, but finding this project amazing and want to make the most of... it could also be a good opportunity to deep dive deeper into LLMs and ultimately be able to contribute to.

**Reply by pmeier** [reply]
> As ragna runs as a forground process, I can't then run further cli commands in the same terminal.
> 
> Have I missed something obvious?

Sorry, I should have been more clear. If you want to run everything from the same terminal, you need to start the worker and REST API as background processes:

```bash
$ ragna worker --num-threads 16 &
$ ragna api --no-start-worker &
$ ragna ui --no-start-api
```

For a `bash`-like shell you just append an ampersand `&` to the command you want to run in the background.

**Reply by NetDevAutomate** [reply]
That’s what I’ve been doing, all good. Just wanted to see if there was a plan to add a demonised option to run this as a background job with job control i.e. ‘ragna start|stop|status’.

**Reply by pmeier** [reply, marked as answer]
Sorry, misunderstood your question in that case. We are not planning anything there. We'll provide a Dockerfile soon and will probably also have `compose.yml` as example to deploy like that. If you want a sneak peek have a look at the [`docker`](https://github.com/Quansight/ragna/tree/docker) branch. That is roughly what we used for our PyData talk. Deploy was actually a lot harder, but I used this for local testing.

**Comment by NetDevAutomate** [comment]
Thank you so much... I'm thinking of running this on a Kubernetes cluster at home to learn more about Kubernetes... so perfect!!!
